rllib/env/vector_env.py

```diff
@@ -232,11 +257,38 @@ def reset_at(self, index: Optional[int] = None) -> EnvObsType:
             index = 0
         return self.envs[index].reset()
 
+    @override(VectorEnv)
+    def restart_at(self, index: Optional[int] = None) -> None:
+        if index is None:
+            index = 0
+        old_env = self.envs[index]
+        # Re-create the sub-env at the new index.
+        self.envs[index] = self.make_env(index)
+        # Try closing down the old (possibly faulty) sub-env, but ignore errors.
+        try:
+            old_env.close()
+        except Exception as e:
+            if log_once("close_sub_env"):
+                logger.warning(
+                    "Trying to close old and replaced sub-environment (at vector "
+                    f"index={index}), but closing resulted in error:\n{e}"
+                )
+
     @override(VectorEnv)
     def vector_step(self, actions):
         obs_batch, rew_batch, done_batch, info_batch = [], [], [], []
         for i in range(self.num_envs):
-            obs, r, done, info = self.envs[i].step(actions[i])
+            try:
+                obs, r, done, info = self.envs[i].step(actions[i])
+            except Exception as e:
+                if self.restart_failed_sub_environments:
+                    self.restart_at(i)
+                    obs = self.envs[i].observation_space.sample()
+                    r = 0.0
+                    done = True
+                    info = {"episode_faulty": True}
+                else:
+                    raise e
```

Comment:
a high level question is should we catch and restart in step() or in Sampler ...
supposedly reset() can also throw exceptions that we need to handle.
the tricky part about catching in Sampler is that then we'd need to know which sub env threw this error.
maybe we should rename restart_failed_sub_environments to try_catch_and_restart_env because that's what is happening here, i.e., if restart_failed_sub_environments=False, the worker would crash and get restarted.

Reply:
I was thinking about this, too, and it would - in theory - be the cleaner solution.
However: Doing this would require our base env API to be better at "bubbling" sub-env errors up to the caller of `poll` and `send_actions`.

So we would have to:
* Catch errors at the call to `poll()` AND at the call to `send_actions` as we won't know, when errors happen.
* Make sure the error information (e.g. which sub-env index failed) is returned from both these calls. Right now, we don't have this information, we assume, everything always runs nicely.

Reply:
Also, if you were to throw an error from `poll` and `send_actions`, such that Sampler would be able to handle the sub-env restarts, then the other information (observations, rewards, ... from the other sub-envs, which may all be ok) would get somewhat lost (or at least we would be in a different stack-track than expected, namely the exception handling track). I feel like this would make things even worse than they are right now.
The way we have it right now is saying:
* Our env adapters are able to capture sub-environment exceptions and handle these as a black-box (the user of the env API will never know other than the "faulty_episode" flag, which it may or may not act on).